src/lib/storage/fsst_segment.cpp

```diff
@@ -0,0 +1,167 @@
+#include "fsst_segment.hpp"
+
+#include <climits>
+#include <sstream>
+#include <string>
+#include "resolve_type.hpp"
+#include "storage/vector_compression/base_compressed_vector.hpp"
+#include "storage/vector_compression/base_vector_decompressor.hpp"
+#include "storage/vector_compression/resolve_compressed_vector_type.hpp"
+#include "utils/assert.hpp"
+#include "utils/performance_warning.hpp"
+
+#include <libfsst.hpp>
+
+namespace opossum {
+
+template <typename T>
+FSSTSegment<T>::FSSTSegment(pmr_vector<unsigned char>& compressed_values,
+                            std::unique_ptr<const BaseCompressedVector>& compressed_offsets,
+                            pmr_vector<uint64_t>& reference_offsets, std::optional<pmr_vector<bool>>& null_values,
+                            uint64_t number_elements_per_reference_bucket, fsst_decoder_t& decoder)
+    : AbstractEncodedSegment{data_type_from_type<pmr_string>()},
+      _compressed_values{std::move(compressed_values)},
+      _compressed_offsets{std::move(compressed_offsets)},
+      _reference_offsets{std::move(reference_offsets)},
+      _null_values{std::move(null_values)},
+      _number_elements_per_reference_bucket{number_elements_per_reference_bucket},
+      _decoder{decoder},
+      _offset_decompressor{_compressed_offsets->create_base_decompressor()} {}
+
+template <typename T>
+AllTypeVariant FSSTSegment<T>::operator[](const ChunkOffset chunk_offset) const {
+  PerformanceWarning("operator[] used");
+  DebugAssert(chunk_offset < size(), "Passed chunk offset must be valid.");
+
+  const auto typed_value = get_typed_value(chunk_offset);
+  if (typed_value) {
+    return typed_value.value();
+  }
+  return NULL_VALUE;
+}
+
+template <typename T>
+uint64_t FSSTSegment<T>::get_offset(const ChunkOffset chunk_offset) const {
+  // For the number of values below 8 (number of reference offsets in encoder) we don't use reference offsets;
```

Comment:
That sounds a bit strange. This extra handling with `== 0` is confusing. Elements of the first chunk will always have the offset of zero, can't we use that directly?

Reply:
Yes, elements of the first chunk will always have an offset of zero, so we can use it directly. 

However, if `_number_elements_per_reference_bucket == 0`, that means, that we don't use reference offsets at all (we do it when the number of offsets is smaller than the number of reference offsets, which is a constant `8`). So, in that case, we can use offsets directly as well, without the need to add reference offsets to them.